dali/operators/segmentation/select_masks.cc

```diff
@@ -0,0 +1,241 @@
+  DALI_ENFORCE(in_polygons_shape.sample_dim() == 2,
+               make_string("``polygons`` input is expected to be 2D. Got ",
+                           in_polygons_shape.sample_dim(), "D"));
+
+  const auto &in_vertices = ws.template InputRef<CPUBackend>(2);
+  auto in_vertices_shape = in_vertices.shape();
+  DALI_ENFORCE(in_vertices_shape.sample_dim() == 2,
+               make_string("``vertices`` input is expected to be 2D. Got ",
+                           in_vertices_shape.sample_dim(), "D"));
+
+  int nsamples = in_polygons.size();
+  DALI_ENFORCE(nsamples == in_polygons_shape.size() && nsamples == in_vertices_shape.size(),
+               make_string("All the inputs should have the same number of samples. Got: ", nsamples,
+                           ", ", in_polygons_shape.size(), ", ", in_vertices_shape.size()));
+
+  if (nsamples == 0) {  // empty input
+    output_desc.reserve(2);
+    output_desc.push_back({in_polygons_shape, in_polygons.type()});
+    output_desc.push_back({in_vertices_shape, in_vertices.type()});
+    return true;
+  }
+
+  for (int i = 0; i < nsamples; i++) {
+    auto sh = in_polygons_shape.tensor_shape_span(i);
+    DALI_ENFORCE(3 == sh[1],
+                 make_string("``polygons`` is expected to contain 2D tensors with 3 columns: "
+                             "``mask_id, start_idx, end_idx``. Got ",
+                             sh[1], " columns."));
+  }
+
+  const auto &in_mask_ids_view = view<const int32_t, 1>(in_mask_ids);
+  const auto &in_polygons_view = view<const int32_t, 2>(in_polygons);
+
+  auto out_polygons_shape = in_polygons_shape;
+  auto out_vertices_shape = in_vertices_shape;
+
+  samples_.resize(nsamples);
+  for (int i = 0; i < nsamples; i++) {
+    samples_[i].clear();
+    auto &selected_masks = samples_[i].selected_masks;
+    auto &polygons = samples_[i].polygons;
+    int64_t nselected = in_mask_ids_view.tensor_shape_span(i)[0];
+    selected_masks = make_cspan(in_mask_ids_view.tensor_data(i), nselected);
+    out_polygons_shape.tensor_shape_span(i)[0] = selected_masks.size();
+    int idx = 0;
+    for (auto mask_id : selected_masks) {
+      if (polygons.find(mask_id) != polygons.end()) {
+        DALI_WARN(make_string("mask_id ", mask_id, " is duplicated. Ignoring..."));
+        continue;
+      }
+      polygons[mask_id].new_mask_id = reindex_masks_ ? idx++ : mask_id;
+    }
+
+    int64_t npolygons = in_polygons_shape.tensor_shape_span(i)[0];
+    int64_t in_nvertices = in_vertices_shape.tensor_shape_span(i)[0];
+    for (int64_t k = 0; k < npolygons; k++) {
+      const auto *poly_data = in_polygons_view.tensor_data(i) + k * 3;
+      int mask_id = poly_data[0];
+      auto it = polygons.find(mask_id);
+      if (it == polygons.end())
+        continue;
+      auto &poly = it->second;
+      poly.start_vertex = poly_data[1];
+      poly.end_vertex = poly_data[2];
+
+      DALI_ENFORCE(
+        poly.start_vertex >= 0 && poly.end_vertex <= in_nvertices,
```

Comment:
Missing check for `poly.start_vertex <= poly.end_vertex`.

Reply:
done